src/module/item/spellcasting-entry/collection.ts

```diff
@@ -144,6 +144,170 @@ export class SpellCollection extends Collection<Embedded<SpellPF2e>> {
+     *  Does nothing if all references are already resolved.
+    async resolveReferences(): Promise<void> {
+        if (!this.actor) return;
+
+        const start = performance.now();
+        let counter = 0;
+        for (const reference of Object.values(this.entry.spellReferences)) {
+            if (!reference.sourceId || this.has(reference._id)) continue;
+
+            const refSpellSource = (await fromUuid<SpellPF2e>(reference.sourceId!))?.toObject();
+            if (!refSpellSource) {
+                console.warn(
+                    `Could not resolve UUID [${reference.sourceId}] in SpellcastingEntry [${this.entry.uuid}]!`
+                );
+                continue;
+            }
+            const refSource = deepClone(reference);
+            delete refSource.sourceId;
+
+            const newSource = mergeObject(refSpellSource, refSource);
+            newSource.system.location.value = this.entry.id;
+            const spell = new SpellPF2e(newSource, { parent: this.actor }) as Embedded<SpellPF2e>;
+            spell.isReference = true;
+
+            this.actor.items.set(spell.id, spell);
+            this.actor.itemTypes.spell.push(spell);
+            this.set(spell.id, spell);
+            counter += 1;
+        }
+        const end = performance.now();
+        console.log(`Execution time (${this.actor.name}) [${this.entry.uuid}] (${counter}): ${end - start} ms`);
+    } */
+
+    /** Resolves spell references and adds derived items to this Collection and the owning Actor.
+     *  Does nothing if all references are already resolved.
+     */
+    async resolveReferences() {
+        if (!this.actor) return;
+        const start = performance.now();
+        const compendiumMap: Map<string, Map<string, SpellReferenceData>> = new Map();
+        for (const reference of Object.values(this.entry.spellReferences)) {
+            if (!reference.sourceId || this.has(reference._id)) continue;
+            const [type, scope, packId, id]: (string | undefined)[] = reference.sourceId.split(".");
+            if (type !== "Compendium") continue;
+
+            const pack = `${scope}.${packId}`;
+            const refSource = deepClone(reference);
+            delete refSource.sourceId;
+
+            if (compendiumMap.has(pack)) {
+                compendiumMap.get(pack)!.set(id, refSource);
+            } else {
+                const idMap: Map<string, SpellReferenceData> = new Map();
+                idMap.set(id, refSource);
+                compendiumMap.set(pack, idMap);
+            }
+        }
+        let counter = 0;
+        for (const [packId, idMap] of compendiumMap.entries()) {
+            const pack = game.packs.get(packId);
+            if (!pack) {
+                console.warn("Pack doesn't exist");
+                continue;
+            }
+            const ids = [...idMap.keys()];
+            const docs = (await pack.getDocuments({ _id: { $in: ids } })) as SpellPF2e[];
+            counter = docs.length;
```

Comment:
+=

Reply:
Counter is a misnomer here I just copied it from the previous method. 😄